python/across_api/base/constraints.py

```diff
@@ -0,0 +1,181 @@
+# Copyright © 2023 United States Government as represented by the
+# Administrator of the National Aeronautics and Space Administration.
+# All Rights Reserved.
+
+from abc import ABC, abstractmethod
+from typing import Optional, Union
+
+import astropy.units as u  # type: ignore[import]
+import numpy as np
+from astropy.coordinates import Angle, SkyCoord  # type: ignore[import]
+from astropy.time import Time  # type: ignore[import]
+from shapely import Polygon, points  # type: ignore[import]
+
+from .ephem import EphemBase
+
+
+def getslice(time: Time, ephem: EphemBase) -> slice:
+    """
+    Return a slice for what the part of the ephemeris that we're using.
+
+    Arguments
+    ---------
+    time
+        The time to calculate the slice for
+    ephem
+        The spacecraft ephemeris
+
+    Returns
+    -------
+        The slice for the ephemeris
+    """
+    # If we're just passing a single time, we can just find the index for that
+    if time.isscalar:
+        # Check that the time is within the ephemeris range, this should never
+        # happen as the ephemeris is generated based on this range.
+        assert (
+            time >= ephem.begin and time <= ephem.end
+        ), "Time outside of ephemeris of range"
+        # Find the index for the time and return a slice for that index
+        index = ephem.ephindex(time)
+        return slice(index, index + 1)
+    else:
+        # Check that the time range is within the ephemeris range, as above.
+        assert (
+            time[0] >= ephem.begin and time[-1] <= ephem.end
+        ), "Time outside of ephemeris of range"
+
+        # Find the indices for the start and end of the time range and return a
+        # slice for that range
+        return slice(ephem.ephindex(time[0]), ephem.ephindex(time[-1]) + 1)
+
+
+class Constraint(ABC):
+    """Define the basic structure of a Constraint class."""
+
+    @abstractmethod
+    def __call__(
+        self, time: Time, ephem: EphemBase, skycoord: Optional[SkyCoord] = None
+    ) -> Union[bool, np.ndarray]:
+        pass
+
+
+class SAAPolygonConstraint(Constraint):
+    """
+    Polygon based SAA constraint. The SAA is defined by a Shapely Polygon, and
+    this constraint will calculate for a given set of times and a given
+    ephemeris whether the spacecraft is in that SAA polygon.
+
+    Attributes
+    ----------
+    polygon
+        Shapely Polygon object defining the SAA polygon.
+    """
+
+    polygon: Polygon
```

Comment:
My only concern is also that this is shapely.Polygon. Should consider re-implementing with spherical reference calculations. 

Maybe starting with an arbitrarily high order healpix map (order=10) and finding the pixels contained by the SAA, and testing whether there is an intersection between the ephem lon lat pixels (in the same order map).

Reply:
I would use Astropy regions. https://astropy-regions.readthedocs.io/en/stable/contains.html

Reply:
I will look into astropy regions, if I can get rid of the shapely dependency I'd be fine with that. 

Regarding the accuracy, I've compared the predicted SAA windows with those predicted by the Swift Flight Operations Team using Satellite ToolKit (STK). They match within seconds, and we're calculating with a 60s time resolution. 

I also happen to have on hand the algorithm that Swift's Flight Software uses to calculate if it's inside the SAA polygon, and it's extremely basic, more so that shapely even, and is certainly not correcting for spherical geometry. 

It's not important to be mathematically perfect, it's important to match what the spacecraft would do. For Swift shapely does. For BurstCube, we'll I can talk with Ava about how they calculate SAA passages so we can best match theirs.

Reply:
It turns out that astropy-regions does _not_ use Shapely to implement geometric predicates, but its implementation of predicates for sky regions _does_ rely on projecting into a WCS and then doing Euclidean geometry. So it would have the same general problems as shapely.

Reply:
I would instead use https://github.com/spacetelescope/spherical_geometry which I have used in the past.

Yes, you'll get the same results in this case because the SAA polygon does not cross the dateline or enclose either pole, but in the likely event that we have to do any _other_ computational geometry on the sphere, we're likely to want to use implementations of spatial predicates that give correct results in general on the sphere.

Reply:
Yes, I just came to the same conclusion that regions isn't fit for this purpose. I'll look into that STScI package.

Reply:
Here is the result of my investigation. I plugged in the `spherical-geometry` library and it works. For Swift side, the results validate using pytest against the Swift API, but then again so does the `shapely` driven calculation also. 

The big issue is that `spherical-geometry` is slow. For single points it's 17x slower than `shapely` and even worse, it does not support vectorization (https://github.com/spacetelescope/spherical_geometry/issues/259), so you have to feed it the points one by one, slowing things down even more. For 86400 data points, calculating SAA using `spherical_geometry` takes 21.4s, for `shapely`, 384ms (M1 MacBook Pro). 

I compared the results against actual SAA passages from Swift using a high resolution (1s) ephemeris, so these are *real numbers* vs prediction. For both the `shapely` and `spherical_geometry` versions of the code, the difference in the calculated SAA entry/exit times vs Swift actual is <<60s, so when we're using 60s time resolution, it's a wash.   

The `spherical_geometry` code is slightly more accurate, but the difference is on the ~2-3s level.

This all leads to my previous conclusion that `shapely` is *good enough* for this purpose, insofar that it matches the spacecraft values fot SAA when calculated at a 60s resolution.

Reply:
I agree, that's not a worthwhile tradeoff. Shapely is an excellent pick for now.

In case we need something like this but more general... I skimmed spherical_geometry and mapped out what it would take to make it vectorize. It's clear that the code was written with that in mind at some level (lots of `np.broadcast_arrays`, the Numpy C code has lots of `PyUFunc_FromFuncAndDataAndSignature` for [generalized ufuncs](https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html)). It wouldn't be a huge lift, but I would guess that it would be a week or two of solid work.

There is also https://github.com/lsst/sphgeom which looks more complete but it would be _mad difficult_ to vectorize effectively because its Python wrapper sits on top of a C++ class library.

Reply:
I will add, that the difference between the calculated SAA using the `shapely` and `spherical-geometry` is more pronounced at low latitudes, so the inaccuracies of the `shapely` method may be more critical for high-inclination orbit missions. 

Attached image shows the difference between the two methods.
<img width="504" alt="Screenshot 2024-02-08 at 20 56 40" src="https://github.com/nasa-gcn/gcn.nasa.gov/assets/15330708/7b739258-445d-4572-aea6-fb575de1e97b">

Reply:
You could minimize those errors by working in a projection (i.e., a WCS) that is centered on the SAA.

Reply:
WCS is a rabbit hole I don't want to go down right now. I will say that just made a HEALPix based SAA constraint, and it's super fast and accurate, faster than `shapely`. I created the HEALPix map using `spherical_geometry` (took 15 mins for a NSIDE=512 map). The created map bzipped is ~3.4kb and is only read in at instantiation of the constraint. This could be a good alternative if we don't just go with the current `shapely` implementation. Basically a lookup table for the `spherical_geometry` version of this constraint.